### DISCUSSION TITLE
After hasing admin password unable to login into the /admin page.

### DISCUSSION THREAD

**Discussion by sebasdt** [dani-garcia/vaultwarden#3454] [Q&A]
### Subject of the issue

After I changed my admin password to a more secure password, my vaultwarden instance is still saying I need to secure it.
This also resulted in me not able to log into the admin page.
### Deployment environment
I'm using portainer and vaultwarden has its own bind volume for saving everything persitantly.

* vaultwarden version:
Latest stable version: 1.28.1

* Install method:
portainer (docker stack)
* Clients used: 

* Reverse proxy and version: 

* MySQL/MariaDB or PostgreSQL version: 

* Other relevant details:
Looking into the vaultwarden volume I dont see a config.json that could conflict with the docker env's
### Steps to reproduce

-     Set up Vaultwarden with plain text admin token
-     After Vaultwarden is up and running with plain text token, generate a hashed token with argon2
-     Change ADMIN_TOKEN to hashed token and restart Vaultwarden
-     Vaultwarden logs warning about using a plain text token with ADMIN_TOKEN set to a hashed token
### Expected behaviour

Im able to log into the admin page and that the logs dont show I need to switch to a more secure password
### Actual behaviour

The /admin page is saying: "Error: Invalid admin token, please try again." and not letting me log in.
### Troubleshooting data

Logs
`
[NOTICE] You are using a plain text `ADMIN_TOKEN` which is insecure.

Please generate a secure Argon2 PHC string by using `vaultwarden hash` or `argon2`.

See: https://github.com/dani-garcia/vaultwarden/wiki/Enabling-admin-page#secure-the-admin_token
`

**docker-compose yml admin env:**
NOTE THIS IS NOT MY TOKEN BUT IT DOES LOOK LIKE THIS!
`
  environment:
    ADMIN_TOKEN: $$argon2id$$v=19$$m=19456,t=2,p=1$$UUZxK1FZMkZoRHFQRlVrTXZvS0E3bHpNQW55c2dBN2NORzdsa0Nxd1JhND0$$cUoId+JBUsJutlG4rfDZayExfjq4TCt48aBc9qsc3UI
`

**Comment by BlackDex** [comment]
Try to put single quotes around it `'`.
Also, make sure you have run `docker compose up -d`, and that you do not have a `config.json` file in your data folder. That will override all environment variables.

**Comment by quyleanh** [comment]
Same issue here.

I have update `docker-compose.yml` file and using `docker compose up -d` command.
However when I checked the `/data/config.json` file. It still shows the old plain text token.
Tried to use this plain text token but it doesn't help.

**Reply by timotheeandres** [reply]
I think you may just delete the line `admin_token=...` from the config file. After I did so, the notice message went away and the token works as expected.

**Reply by quyleanh** [reply]
I did try it and I can access to admin page. But whenever I set up token, it will throw 401 error.

**Reply by timotheeandres** [reply]
What do you mean by `whenever I set up token`? Do you try to set a new one up inside of the admin panel?

**Reply by quyleanh** [reply]
Yes. I tried to set new one inside the admin panel and it has error when logging in.
You can refer to this discussion #3458

**Reply by BlackDex** [reply]
Ok, and the one you setup inside the admin settings interface, is that with or without double dollar signs `$$`?
Because there it needs to be **without** `$$`, just a single `$` **and** no `'`/`"` quotes surrounding that value!

**Reply by quyleanh** [reply]
When setting token in the admin page, I use single `$` and without `'` or `"`.

**Comment by EricDriussi** [comment]
Hi there lovely people! I think I'm having the same issue and in my case it's pretty easy to reproduce:

- Create the token:

```sh
echo -n "MySecretPassword" | argon2 "$(openssl rand -base64 32)" -e -id -k 19456 -t 2 -p 1
```

- Run the docker container (for some reason it doesn't work if I don't point to a persisten volume):

```sh
docker run -d -e ADMIN_TOKEN='$argon2id$v=19$m=19456,t=2,p=1$NzFtWlRHZTR5VG1Dc3JFK3d6RmFXbGwxY0xQNHpxZXNYcGRIOHJNY1N4ND0$TyEimYSycuTs5/Ad0XhjutX8kB08qR5QNhDwFqAoAeY' -p 8080:80 -v ~/test/vault:/data/ vaultwarden/server:latest
```

- Go to `http://localhost:8080/admin` and paste the admin token:

```
$argon2id$v=19$m=19456,t=2,p=1$NzFtWlRHZTR5VG1Dc3JFK3d6RmFXbGwxY0xQNHpxZXNYcGRIOHJNY1N4ND0$TyEimYSycuTs5/Ad0XhjutX8kB08qR5QNhDwFqAoAeY
```

- Get `401` :(

![image](https://user-images.githubusercontent.com/46979145/235345319-e15ae980-619a-4e98-92a6-50c84572f872.png)

Bear in mind that, to ensure I'm isolating the issue, I'm nuking all docker containers and volumes prior to running the above mentioned commands:

```sh
docker stop $(docker container ls -a -q); docker system prune -a -f --volumes
```

I understand what has been mentioned here regarding quotes and special chars like `$` but that really doesn't seem to be the issue for me (at least as far as I can tell).

Oh and to be clear, using a plain text token works without any issues, it only seem to affect the `argon2` token.

Thanks in advance and let me know If there's any other information I can provide!

**Reply by kpfleming** [reply]
You're not using a 'persistent volume', you're binding a directory from your filesystem (`~/test/vault`) into the container. `docker prune` isn't going to empty the contents of that directory, you'll have to do it directly if that's your goal.

**Reply by EricDriussi** [reply]
I would say that binding a dir to the container makes the data persistent (as it lives beyond the run of the container) but you are right in that a prune won't clear the binded dir.
I am however removing it manually to ensure each run is unaffected by the prior one (which I forgot to mention).
Thanks for pointing that out!

**Reply by BlackDex** [reply]
I'm running it like this:
```bash
docker run --rm -e ADMIN_TOKEN='$argon2id$v=19$m=19456,t=2,p=1$NzFtWlRHZTR5VG1Dc3JFK3d6RmFXbGwxY0xQNHpxZXNYcGRIOHJNY1N4ND0$TyEimYSycuTs5/Ad0XhjutX8kB08qR5QNhDwFqAoAeY' -e I_REALLY_WANT_VOLATILE_STORAGE=true -p8080:80 vaultwarden/server:latest
```

Works just fine for me.
Using the generated hash from your post.
Either there still is a `config.json` which holds a different password, or something else strange is happening for you. But running it like above works just fine 😄

**Reply by EricDriussi** [reply]
Welp, turns out I **_though_** I was removing the linked dir between runs...
That `I_REALLY_WANT_VOLATILE_STORAGE=true` kinda solved the issue :man_facepalming: 

Thanks a lot for your time and pardon my lack of brain cells :face_with_spiral_eyes:

**Comment by gustavohellwig** [comment]
I had the same issue.
Solution:
`cd /opt/vaultwardenasexample/
sudo docker-compose down
mv yourvwdata/config.json yourvwdata/config-old.json
sudo docker-compose up -d
diff yourvwdata/config.json yourvwdata/config-old.json 
`
Login by the /admin and press save. After that, compare your old file. 
`
diff yourvwdata/config.json yourvwdata/config-old.json 
`
If it's all ok, remove the old one.

**Comment by Hundhausen** [comment]
I've noticed that someone in the line had an issue with my used Password. The length was not an issue but apparently a special character. Here are my used special characters `#\&:.(?[}{^$)|=` (removed duplicates and alphanumeric characters). At no point someone threw an error, except when trying to login. When removing the special characters and doing the same steps, it worked on the first try.

The length alone should be save enough but I like to go overkill but not everyone supports everything. For me it sounds like a security issue, when a character causes an issue. Might be not exploitable (could be argon2 or ubuntu that removed or altered characters) but surely something that someone should look into and either fix it or clarifying in the docs, what chars aren't supported or which are supported. Don't know if emojis in passwords are a thing but when you use the whole Unicode range (or need to expect it), you have a big entropy, when I'm not mistaken.

**Reply by BlackDex** [reply]
Could you post how you generated the Argon2 string?

**Comment by Hundhausen** [comment]
> Could you post how you generated the Argon2 string?

I've used Ubuntu 22.04 via WSL and SSH and thats the command: `echo -n "#\&:.(?[}{^$)|=" | argon2 "$(openssl rand -base64 32)" -e -id -k 65540 -t 3 -p 4`
I've used a different password and just included the special chars mentioned before. The result looked like it should be.

I don't know if this is relevant but I had one character double `##`.

**Reply by BlackDex** [reply]
What happens if you use single quotes `'`

**Reply by Hundhausen** [reply]
Single quotes work. So no big issue.

**Reply by BlackDex** [reply]
I mean surrounding those special characters 😉

**Reply by BlackDex** [reply]
I don't know from the top of my head if that has some impact in the end results

**Reply by Hundhausen** [reply]
Oh yeah, I did that. I swapped both `"` to `'` and it worked. I even tested Emojis and other special characters and it worked without an issue. Adding a single quote into the password, would not result with a token and a double quote does display a token but using this  and the password doesn't work.

**Comment by ShlomiD83** [comment]
Hi, I didn't quite understand why I see in the logs a message that I'm using an admin token in plain text.
I've used the Vaultwarden hash without ' but with $$ as the documentation suggested.
Tried different methods as well (single $ with ' and without).
I'm able to connect to the admin page without any issues. 
Why am I seeing that message?

**Reply by BlackDex** [reply]
Where did you configured the token? Because if you configured it as an environment variable, but got also have a config.json file, then ots overruled by it. Set the token via the admin interface without quotes and just a single `$`, that should solve your issue.

**Reply by ShlomiD83** [reply]
so I should remove the admin token from the env variable? or remove config.json?

**Comment by ShlomiD83** [comment]
> Where did you configured the token? Because if you configured it as an environment variable, but got also have a config.json file, then ots overruled by it. Set the token via the admin interface without quotes and just a single `$`, that should solve your issue.

I've commented out the admin token from my portainer stack and configured it as you suggested and it seems that the message is gone.
thanks again for all your help & hard work.

**Comment by http403** [comment]
Same issues but in Pikapods.

I have the token set via the `/admin` endpoint as Pikapods doesn't allow the encoded hash into its interface. I had verified that the hash in `config.json` is the correct one but somehow doesn't let me in.

**Reply by BlackDex** [reply]
What does the Vaultwarden log say during startup?
Also, do not add it there with quotes or the double `$$`, that will fail.

**Reply by http403** [reply]
I found what I did wrong. I forgot to specify `-n` with `echo` and causing the problem. Sorry for the false alarm. Now it is working perfectly.

**Reply by BlackDex** [reply]
Ah good that you found the issue, and thanks for sharing.
And np.
